### PR TITLE
mountinfo: more docs

### PR DESCRIPTION
**Pull request by kolyshkin**
Closes: #14

### CODE REVIEW

**Comment by kolyshkin** [comment]
@thaJeztah PTAL

**Comment by kolyshkin** [comment]
@cpuguy83 PTAL

File: mountinfo/doc.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// see different mounts. A per-process mountinfo table is available from /proc/<PID>/mountinfo,
+// where <PID> is a numerical process identifier.
+//
+// In general, /proc is not a very effective interface, and mountinfo is not an exception.
+// For example, there is no way to get information about a specific mount point (i.e. it
+// is all-or-nothing). This package tries to hide the /proc ineffectiveness by using
+// parse filters while reading mountinfo. A filter can skip some entries, or stop
+// processing the rest of the file once the needed information is found.
+//
+// For any functions that have path as an argument (such as Mounted or various filters),
```
Comment (thaJeztah) [inline mountinfo/doc.go:17]:
```suggestion
// For functions that have path as an argument (such as Mounted or various filters),
```

File: mountinfo/doc.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// the argument must be an absolute path, with all symlinks resolved, and cleaned.
```
Comment (thaJeztah) [inline mountinfo/doc.go:18]:
```suggestion
// the argument must be an absolute path, with all symlinks resolved and cleaned.
```

File: mountinfo/doc.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// One way to achieve this is to process the path using filepath.EvalSymlinks.
+//
+// NOTE that in many cases there is no need to consult mountinfo at all. Here are some
+// of the cases where mountinfo should not be parsed:
+//
+// 1. Before performing a mount. If you want to check whether a directory is mounted,
+//    call os.Lstat on it and its parent directory, and compare their
+//    st.Sys().(*syscall.Stat_t).Dev fields -- if they differ, then the directory is
+//    the mount point. Optionally, you can also call unix.Statfs and check the Type
+//    field (i.e. filesystem type).
```
Comment (thaJeztah) [inline mountinfo/doc.go:28]: Would it be useful to have utilities for this? Or too hard to have some thing that's universal enough?

File: mountinfo/doc.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+//
+// 2. After performing a mount. If there is no error returned, the mount succeded;
```
Comment (thaJeztah) [inline mountinfo/doc.go:30]:
```suggestion
// 2. After performing a mount. If there is no error returned, the mount succeeded;
```

File: mountinfo/mountinfo.go [hunk 5]
```diff
@@ -18,6 +18,10 @@ func GetMountsFromReader(reader io.Reader, f FilterFunc) ([]*Info, error) {
 
 // Mounted determines if a specified mountpoint has been mounted.
 // On Linux it looks at /proc/self/mountinfo.
+//
+// The argument must be an absolute path, with all symlinks resolved, and clean.
```
Comment (thaJeztah) [inline mountinfo/mountinfo.go:22]:
```suggestion
// The argument must be an absolute path, with all symlinks resolved and cleaned.
```

File: mountinfo/mountinfo.go [hunk 6]
```diff
@@ -18,6 +18,10 @@ func GetMountsFromReader(reader io.Reader, f FilterFunc) ([]*Info, error) {
...
+// One way to ensure it is to process the path using filepath.EvalSymlinks before
```
Comment (thaJeztah) [inline mountinfo/mountinfo.go:23]: `it` is a bit ambiguous here, perhaps combine with the previous sentence?

`for example, using filepath.EvalSymlinks()`

File: mountinfo/doc.go (hunk 2 above)
Comment (kolyshkin) [reply mountinfo/doc.go:18]: the "clean" part is not about symlinks, but about whatever filepath.Clean() does. If we drop a comma here, it sounds like "cleaned" is about symlinks.

File: mountinfo/doc.go (hunk 3 above)
Comment (kolyshkin) [reply mountinfo/doc.go:28]: I actually plan to change Mounted() to use the above algo, similar to https://github.com/cilium/cilium/pull/12249.

Let me actually do this...

Comment (kolyshkin) [reply mountinfo/doc.go:28]: https://github.com/moby/sys/pull/20

If we'll merge that one before this, I'll change the text here.

File: mountinfo/doc.go (hunk 4 above)
Comment (kolyshkin) [reply mountinfo/doc.go:30]: fixed

File: mountinfo/doc.go (hunk 1 above)
Comment (kolyshkin) [reply mountinfo/doc.go:17]: fixed

File: mountinfo/mountinfo.go (hunk 6 above)
Comment (kolyshkin) [reply mountinfo/mountinfo.go:23]: I have worded it a bit differently

File: mountinfo/doc.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+//
+// 2. After performing a mount. If there is no error returned, the mount succeeded;
+//    checking the mount table for a new mount is redundant and expensive.
+//
+// 3. Before performing an unmount. It is more efficient to do an unmount and ignore
+//    a specific error (EINVAL) which tells the directory is not mounted.
+//
+// 4. After performing an unmount. If there is no error returned, the unmount succeeded.
+//
+// 5. To find the mount point root of a specific directory. You can perform os.Stat()
```
Comment (cpuguy83) [inline mountinfo/doc.go:44]: Nice tip!

Comment (kolyshkin) [reply mountinfo/doc.go:44]: Using this in runc; will probably add a function to do that in here.

File: mountinfo/doc.go (hunk 3 above)
Comment (thaJeztah) [reply mountinfo/doc.go]: Thanks!

File: mountinfo/doc.go (hunk 2 above)
Comment (thaJeztah) [reply mountinfo/doc.go]: Ah, yes, makes sense

**Pull request merged**